src/revwalk.c

```diff
@@ -398,81 +398,191 @@ static int revwalk_next_reverse(git_commit_list_node **object_out, git_revwalk *
 	return *object_out ? 0 : GIT_ITEROVER;
 }
 
-
-static int interesting(git_pqueue *list)
+static int contains(git_pqueue *list, git_commit_list_node *node)
 {
 	size_t i;
 
 	for (i = 0; i < git_pqueue_size(list); i++) {
 		git_commit_list_node *commit = git_pqueue_get(list, i);
-		if (!commit->uninteresting)
+		if (commit == node)
 			return 1;
 	}
 
 	return 0;
 }
 
-static int contains(git_pqueue *list, git_commit_list_node *node)
+static void mark_parents_uninteresting(git_commit_list_node *commit)
 {
-	size_t i;
+	unsigned short i;
+	git_commit_list *parents = NULL;
 
-	for (i = 0; i < git_pqueue_size(list); i++) {
-		git_commit_list_node *commit = git_pqueue_get(list, i);
-		if (commit == node)
-			return 1;
-	}
+	for (i = 0; i < commit->out_degree; i++)
+		git_commit_list_insert(commit->parents[i], &parents);
 
-	return 0;
+
+	while (parents) {
+		git_commit_list_node *commit = git_commit_list_pop(&parents);
+
+		while (commit) {
```

Comment:
Is this second loop walking up the first-parent chain even necessary? We add the commit's parents to `parents` either way and would arrive at all parents either way. Additionally, I think we iterate over all first-parents twice due to this (even though the second time we bail out immediately due to `commit->uninteresting == 1`).